state.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"io"
+	"time"
+
+	"github.com/mendersoftware/log"
+	"github.com/pkg/errors"
+)
+
+type State interface {
+	// Perform state action, returns next state and boolean flag indicating if
+	// execution was cancelled or not
+	Handle(c Controler) (State, bool)
+	// Cancel state action, returns true if action was cancelled
+	Cancel() bool
+	// Return numeric state ID
+	Id() MenderState
+}
+
+type StateRunner interface {
+	// Set runner's state to 's'
+	SetState(s State)
+	// Obtain runner's state
+	GetState() State
+	// TODO generic state run action
+}
+
+var (
+	initState = &InitState{
+		BaseState{
+			id: MenderStateInit,
+		},
+	}
+
+	bootstrappedState = &BootstrappedState{
+		BaseState{
+			id: MenderStateBootstrapped,
+		},
+	}
+
+	updateCheckWaitState = NewUpdateCheckWaitState()
+
+	updateCheckState = &UpdateCheckState{
+		BaseState{
+			id: MenderStateUpdateCheck,
+		},
+	}
+
+	updateCommitState = &UpdateCommitState{
+		BaseState{
+			id: MenderStateUpdateCommit,
+		},
+	}
+
+	rebootState = &RebootState{
+		BaseState{
+			id: MenderStateReboot,
+		},
+	}
+
+	doneState = &FinalState{
+		BaseState{
+			id: MenderStateDone,
+		},
+	}
+)
+
+// Helper base state with some convenience methods
+type BaseState struct {
+	id MenderState
```

Comment:
This is more internal thing, so it doesn't matter that much, but maybe we can have `state` here as well?